scalameta/quasiquotes/shared/src/main/scala-3/scala/meta/internal/quasiquotes/ExprLifts.scala

```diff
@@ -0,0 +1,542 @@
+package scala.meta
+package internal
+package quasiquotes
+
+import scala.runtime.ScalaRunTime
+import scala.quoted._
+import scala.meta.{Tree => MetaTree}
+import scala.meta.internal.trees.Quasi
+import scala.meta.internal.parsers.Absolutize._
+import scala.meta.internal.parsers.Messages
+
+import scala.collection.mutable
+import scala.annotation.tailrec
+import scala.meta.trees.Origin
+
+// Uses holes in implementation, for which we do not have access here, so we implement those elsewhere
+trait InternalTrait(using val internalQuotes: Quotes) {
+  def liftTree(tree: MetaTree): internalQuotes.reflect.Tree
+  def liftOptionTree[T: Type](maybeTree: Option[MetaTree]): internalQuotes.reflect.Tree
+  def liftTrees[T: Type](trees: Seq[MetaTree]): internalQuotes.reflect.Tree
+  def liftTreess(treess: List[List[MetaTree]]): internalQuotes.reflect.Tree
+  def liftQuasi0(quasi: Quasi, optional: Boolean = false): internalQuotes.reflect.Tree
+  def liftOrigin(origin: Origin): internalQuotes.reflect.Tree
+
+  protected def unquotesName(q: scala.meta.internal.trees.Quasi): Boolean
+}
+
+// In Scala 2 version, this part of the codebase was automatically generated by a macro,
+// with materializeAst[MetaTree].
+// Probably would be better if it was also generated here, by a similar macro or by a custom codegen script.
```

Comment:
probably would be better. since we keep adding trees, this code would have to continue evolving, and it might become high-maintenance. you probably explained why it was hard or impossible to do, i need to go over your earlier comments.